### ISSUE TITLE
Question regarding recent changes

### ISSUE THREAD

**Issue by kubagalecki** [ccfd/courses#85]
@llaniewski I wanted to ask for your input regarding the changes I made today (unfortunately couldn't wait for your approval under a PR due to time pressure). Specifically:
- I migrated the CI from Travis to GitHub actions (Travis was no longer working - I think it's no longer free?). Is this an acceptable solution for you? Furthermore, this opens up the door to some improvements, e.g., caching the tooling used to build the site in a Docker image, so that deployment to gh-pages takes seconds instead of 11 minutes. Do you think it's worth the trouble, or are things fine as is?
- I enabled branch protection, so that now a PR and min. 1 review is required for changes to be published to gh-pages. Seems to me like the previous settings were a bit too lax, it was too easy for someone to break something by accident. Thoughts?

I appreciate your time, cheers.

**Comment by kubagalecki** [comment]
@llaniewski thanks for getting back to me so quickly, you raise some good points.
- [ ] I was thinking of doing something similar to what [I do in my project](https://github.com/kubagalecki/L3STER/actions/runs/1269985708) (based on [this action](https://github.com/marketplace/actions/build-docker-images-using-cache)). Basically, we just store the image on DockerHub, and if any changes to the Dockerfile in the repo are detected, the image is rebuilt. To rebuild the image used in CI, we can simply delete the one on DockerHub. If you find this solution acceptable, I'll implement it when I have some time.
- [ ] This is a very good idea, I'll get right on it. GitHub Actions is a little bit annoying when it comes to exporting/sharing data between jobs, so I think the simplest solution will be to just build the page twice: when the PR is created (the html files can be downloaded and viewed by the reviewer) and when it's merged. The downside is that until I get around to implementing the idea from the previous bullet, the 2 builds will take a long time. Though to be honest, I'm having a hard time imagining an emergency situation involving lab instructions :sweat_smile:

**Comment by ggruszczynski** [comment]
@kubagalecki 
```
I enabled branch protection, so that now a PR and min. 
1 review is required for changes to be published to gh-pages. 
Seems to me like the previous settings were a bit too lax, it was too easy for someone to break something by accident. Thoughts?
```

IMHO it would be enough to have `passed` the PR build to merge.

I think that the 2nd build after the merge is done deliberately by GitHub.

**Comment by llaniewski** [comment]
@kubagalecki Both points sound great. Good to see somebody is keeping this alive, as I think this page is a good consistent approach for lab instructions.

@ggruszczynski @kubagalecki As for the reviews, I think it's a good rule, but I'm worried that the reality of academic work will verify it's applicability :-) But I think it's a good think to at least try :+1: 

Sidenote: I have to switch TCLB from Travis-CI, but I constantly lack the time.

**Comment by kubagalecki** [comment]
I've implemented build checks in #86 - you can now download the raw html files from the PR browse them locally. Would love a review.

@ggruszczynski The 2 builds are triggered by 2 separate actions - this is how I intended it to work. The more optimal solution would be to cache the build from the PR and simply deploy it in the action triggered post-merge. However, I'm not convinced this would work correctly with multiple simultaneous PRs.

As for requiring reviews - I'm open to discussing this, I don't really have strong feelings either way. This can be changed in the branch protection rules for `master`.

@llaniewski I think it's important to keep it alive, since this is the main source of lab instructions for literally hundreds of students. Hopefully GitHub actions will prove more stable than Travis in the long run. If you want any tips for migrating from Travis, my DMs are open :innocent:
